src/python/pants/backend/go/goals/debug_goals.py

```diff
@@ -114,6 +121,74 @@ async def dump_go_import_paths_for_module(
     return DumpGoImportPathsForModule(exit_code=0)
 
 
+class GoExportCgoCodegenSubsystem(GoalSubsystem):
+    name = "go-export-cgo-codegen"
+    help = "Export files generated by Cgo."
+
+
+class GoExportCgoCodegen(Goal):
+    subsystem_cls = GoExportCgoCodegenSubsystem
+
+
+@goal_rule
+async def go_export_cgo_codegen(
+    targets: Targets,
+    console: Console,
+    distdir_path: DistDir,
+    workspace: Workspace,
+    golang_subsystem: GolangSubsystem,
+) -> GoExportCgoCodegen:
+    if not golang_subsystem.cgo_enabled:
+        raise ValueError("Nothing to export since cgo is disabled.")
+
+    go_package_targets = [tgt for tgt in targets if tgt.has_field(GoPackageSourcesField)]
+    for tgt in go_package_targets:
+        # Analyze the package and ensure it is actually contains cgo code.
+        analysis_wrapper = await Get(
+            FallibleFirstPartyPkgAnalysis, FirstPartyPkgAnalysisRequest(tgt.address)
+        )
+        if not analysis_wrapper.analysis:
+            console.write_stdout(
+                f"{tgt.address}: Failed to analyze target: {analysis_wrapper.stderr}\n"
+            )
+            continue
+        analysis = analysis_wrapper.analysis
+        if not analysis.cgo_files:
+            console.write_stdout(
+                f"{tgt.address}: Nothing to export because package does not contain Cgo files.\n"
+            )
+            continue
+        fallible_digest_info = await Get(
+            FallibleFirstPartyPkgDigest, FirstPartyPkgDigestRequest(tgt.address)
+        )
+        if not fallible_digest_info.pkg_digest:
+            console.write_stdout(
+                f"{tgt.address}: Failed to export due to failure to obtain package digest: {fallible_digest_info.stderr}\n"
+            )
+            continue
+
+        # Perform CGo compilation.
+        result = await Get(
+            CGoCompileResult,
+            CGoCompileRequest(
+                import_path=analysis.import_path,
+                pkg_name=analysis.name,
+                digest=fallible_digest_info.pkg_digest.digest,
+                dir_path=analysis.dir_path,
+                cgo_files=analysis.cgo_files,
+                cgo_flags=analysis.cgo_flags,
+            ),
+        )
+
+        output_digest = await Get(Digest, RemovePrefix(result.digest, analysis.dir_path))
+
+        export_dir = distdir_path.relpath / "__cgo__" / tgt.address.path_safe_spec
```

Comment:
Why the dunders? I could be wrong but I don't think we do this sort of thing in other export subdirs? these are paths that end users poke at, so we probably want them to be friendly?

Reply:
Since the cgo codegen is really an internal detail, I probably used the dunders with "internal" in mind. I don't particularly care what the prefix is. Suggestions?

Reply:
Well, but with `export` the user is asking to expose internals, like with codegen. So maybe just `cgo` ?

Reply:
Done.